### PR TITLE
Add a cookiecutter based tutorial to build a custom widget

### LINKED ISSUE #2731: Cookiecutter based tutorial
The previous "Custom Widget" has now been removed from the documentation (see https://github.com/jupyter-widgets/ipywidgets/pull/2677), as it was using patterns specific to the classic notebook.

As mentioned in [this comment](https://github.com/jupyter-widgets/ipywidgets/pull/2677#issue-360008156), we could instead have a tutorial based on the [TypeScript tutorial](https://github.com/jupyter-widgets/widget-ts-cookiecutter), similar to the [Let’s Make an Astronomy Picture of the Day JupyterLab Extension](https://jupyterlab.readthedocs.io/en/stable/developer/extension_tutorial.html).

### PR DESCRIPTION
**Pull request by jtpio**
Fixes #2731 

The idea is to update the tutorial so it uses the TypeScript cookiecutter to bootstrap the widget.

For now we can probably keep the content and narration of the original tutorial (email widget, screenshot in https://github.com/jupyter-widgets/ipywidgets/issues/2731#issuecomment-656121946), and we might want to iterate on it more later.

Structure and narration inspired by: https://jupyterlab.readthedocs.io/en/latest/developer/extension_tutorial.html

**Screenshot:**

<details>

Screenshot_2020-07-14 Building a Custom Widget - Email widget — Jupyter Widgets 8 0 0a0 documentation

</details>

### TODO

- [x] Add doc for the cookiecutter based setup
- [x] Update the instructions so they match the cookiecutter structure ("modify this files with the following content")
- [x] Post a screenshot of the rendered page
- [x] Add screenshot

### CODE REVIEW

**Comment by jtpio** [comment]
This should now cover the content of the previous tutorial and be ready for review.

**Review by ianhi — COMMENTED** [review]
Definitely a helpful read. I left a few comments on some things that might be nice to mention. 

Also: I think it would be great to have a screenshot of the finished product right at the top as a hook to capture interest. The widget architecture screenshot is useful but it is not the most inviting start to the tutorial.

**Comment by jtpio** [comment]
> I think it would be great to have a screenshot of the finished product right at the top as a hook to capture interest

Good idea, added that to the todo list :+1:

File: docs/source/examples/Widget Custom.ipynb [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+    "\n",
+    "When prompted, enter the desired values as follows:\n",
+    "\n",
+    "```bash\n",
+    "author_name []: Your Name\n",
+    "author_email []: your@name.net\n",
+    "github_project_name []: ipyemail\n",
+    "github_organization_name []: \n",
+    "python_package_name [ipyemail]:\n",
+    "npm_package_name [ipyemail]:\n",
```
[the remark this answers is not in the document]
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:96]: Looks like we could indeed make that change if that's the convention to follow :+1:

File: docs/source/examples/Widget Custom.ipynb [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+    "\n",
+    "Both the `DOMWidget` and `ValueWidget` classes inherit from the `Widget` class. The `Widget` class is useful for cases in which the widget is not meant to be displayed directly in the notebook, but instead as a child of another rendering environment.  Here are some examples:\n",
+    "\n",
+    "- If you wanted to create a [three.js](https://threejs.org/) widget (three.js is a popular WebGL library), you would implement the rendering window as a `DOMWidget` and any 3D objects or lights meant to be rendered in that window as `Widget`\n",
+    "- If you wanted to create a widget that displays directly in the notebook for usage with `interact` (like `IntSlider`), you should multiple inherit from both `DOMWidget` and `ValueWidget`. \n",
+    "- If you wanted to create a widget that provides a value to `interact` but is controlled and viewed by another widget or an external source, you should inherit from only `ValueWidget`"
```
[the remark this answers is not in the document]
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:206]: This part was borrowed from the previous tutorial, but we could rephrase it for more clarity.

**Comment by MicaelJarniac** [comment]
There are references to `Widget Custom.ipynb` on other files (#3061). At least one of those references includes a specific anchor (`#DOMWidget,-ValueWidget-and-Widget`), which may need updating.

**Comment by jtpio** [comment]
Thanks @MicaelJarniac for the pointer.

Adding the PR to the 8.0 milestone since the previous notebook-based tutorial was removed in https://github.com/jupyter-widgets/ipywidgets/pull/2677

File: docs/source/examples/Widget Custom.ipynb [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+   },
+   "source": [
+    "### sync=True traitlets"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Traitlets is an IPython library for defining type-safe properties on configurable objects.  For this tutorial you do not need to worry about the *configurable* piece of the traitlets machinery.   The `sync=True` keyword argument tells the widget framework to handle synchronizing that value to the browser.  Without `sync=True`, attributes of the widget won't be synchronized with the front-end."
```
[the remark this answers is not in the document]
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb]: Thanks :+1: 

Done in https://github.com/jupyter-widgets/ipywidgets/pull/2919/commits/4e34ada06d04e8a5ad5d3659c7632acab73cb433

**Comment by jtpio** [comment]
Add a screenshot for the end result:

end-result

File: docs/source/examples/Widget Custom.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb]: Fixed.

File: docs/source/examples/Widget Custom.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:206]: I think "external source" refers to something that does not result from interacting with a DOM element.

[the remark this answers is not in the document]
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:206]: How do you think we could we rephrase this to make it clearer?

**Comment by jtpio** [comment]
@ianhi I've pushed a couple of changes based on the comments above, thanks!

Ideally we will enable the read the docs autobuild for pull requests soon: https://github.com/jupyter-widgets/ipywidgets/issues/2921

So the tutorial can be reviewed more easily.

**Comment by jtpio** [comment]
Also, we will want to update the tutorial for JupyterLab 3.0 too, when changes are backported from the `7.x` branch to `master`.

This can probably be tracked and done in a separate PR.

**Comment by ianhi** [comment]
I'm going to close and re-open to try to pick up the RTD build on PRs

File: docs/source/examples/Widget Custom.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ianhi) [reply docs/source/examples/Widget Custom.ipynb:206]: How about: 

```suggestion
    "- If you wanted to create a widget that provides a value to `interact` but does not need to be displayed, you should inherit from only `ValueWidget`"
```

Also maybe all of these should have `wanted -> want`

[the remark this answers is not in the document]
Comment (ianhi) [reply docs/source/examples/Widget Custom.ipynb:206]: Also these bullet points repeat a lot of info from the paragraphs directly above them. Ideally they would integrated.

File: docs/source/examples/Widget Custom.ipynb [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+   "metadata": {},
+   "source": [
+    "### DOMWidget, ValueWidget and Widget\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To define a widget, you must inherit from the `DOMWidget`, `ValueWidget`, or `Widget` base class.  If you intend for your widget to be displayed, you'll want to inherit from `DOMWidget`.  If you intend for your widget to be used as an input for [interact](./Using%20Interact.ipynb), you'll want to inherit from `ValueWidget`. Your widget should inherit from `ValueWidget` if it has a single ovious output (for example, the output of an `IntSlider` is clearly the current value of the slider).\n",
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:200]:
```suggestion
    "To define a widget, you must inherit from the `DOMWidget`, `ValueWidget`, or `Widget` base class.  If you intend for your widget to be displayed, you'll want to inherit from `DOMWidget`.  If you intend for your widget to be used as an input for [interact](./Using%20Interact.ipynb), you'll want to inherit from `ValueWidget`. Your widget should inherit from `ValueWidget` if it has a single obvious output (for example, the output of an `IntSlider` is clearly the current value of the slider).\n",
```

File: docs/source/examples/Widget Custom.ipynb [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In `ipyemail/__init__.py`, change the import from:\n",
+    "\n",
+    "```python\n",
+    "from .widget import ExampleWidget\n",
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:258]: This is the actual text in the file:
```suggestion
    "from .example import ExampleWidget\n",
```

File: docs/source/examples/Widget Custom.ipynb [hunk 6]
```diff
@@ -0,0 +1,785 @@
...
+    "\n",
+    "A handle to the widget's default DOM element can be acquired via `this.el`.  The `el` property is the DOM element associated with the view.\n",
+    "\n",
+    "In `src/widget.ts`, define the `_emailInput` attribute:\n",
+    "\n",
+    "```typescript\n",
+    "export class EmailView extends DOMWidgetView {\n",
+    "  private _emailInput: HTMLInputElement;\n",
+    "}\n",
+    "```\n",
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:447]: I found this confusing. It made it seem that that there shouldn't be anything in the brackets, i.e. delete the render method.

Perhaps better would be
```
export class EmailView extends DOMWidgetView {
  private _emailInput: HTMLInputElement;

  render() {
     .....
}

File: docs/source/examples/Widget Custom.ipynb [hunk 7]
```diff
@@ -0,0 +1,785 @@
...
+    "\n",
+    "Then, add the following logic for the `render` method:\n",
+    "\n",
+    "```typescript\n",
+    "render() { \n",
+    "    this._emailInput = document.createElement('input');\n",
+    "    this._emailInput.type = 'email';\n",
+    "    this._emailInput.value = 'example@example.com';\n",
+    "    this._emailInput.disabled = true;\n",
+    "    this.el.appendChild(this._emailInput);\n",
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:457]: Similar here to above comment. Is the intention for the entire render method to look like this, or should these lines be added to the line from the cookiecutter?:
```
    this.el.classList.add('custom-widget');

    this.value_changed();
    this.model.on('change:value', this.value_changed, this);
```

File: docs/source/examples/Widget Custom.ipynb [hunk 8]
```diff
@@ -0,0 +1,785 @@
...
+    "@register\n",
+    "class Email(DOMWidget, ValueWidget):\n",
+    "    _model_name = Unicode('EmailModel').tag(sync=True)\n",
+    "    _model_module = Unicode(module_name).tag(sync=True)\n",
+    "    _model_module_version = Unicode(module_version).tag(sync=True)\n",
+    "\n",
+    "    _view_name = Unicode('EmailView').tag(sync=True)\n",
+    "    _view_module = Unicode(module_name).tag(sync=True)\n",
+    "    _view_module_version = Unicode(module_version).tag(sync=True)\n",
+    "```"
```
[the remark this answers is not in the document]
Comment (ianhi) [reply docs/source/examples/Widget Custom.ipynb:248]: Having followed the tutorial further down I think there should  probably also be a `disabled = Bool(False).tag(sync=True)` attribute.

File: docs/source/examples/Widget Custom.ipynb [hunk 9]
```diff
@@ -0,0 +1,785 @@
...
+    "  private _emailInput: HTMLInputElement;\n",
+    "}\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This allows us to update the value from the Python kernel to the views. Now to get the value updated from the front-end to the Python kernel (when the input is not disabled) we can do it using the `model.set` method.\n",
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:654]: Need to also mention `model.save_changes`. I'm pretty sure the difference is that `set` will set the ts model and all thew views will be synced together, and then on `save_changes` the python side will updated.

Comment (ianhi) [reply docs/source/examples/Widget Custom.ipynb:654]:
```suggestion
    "This allows us to update the value from the Python kernel to the views. Now to get the value updated from the front-end to the Python kernel (when the input is not disabled) we set the value on the frontend model using `model.set` and then sync the frontend model with the Python object using `model.save_changes`.\n",
```

File: docs/source/examples/Widget Custom.ipynb [hunk 10]
```diff
@@ -0,0 +1,785 @@
...
+   "metadata": {},
+   "source": [
+    "### Models and views"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The IPython widget framework front end relies heavily on [Backbone.js](http://backbonejs.org/).  Backbone.js is an MVC (model view controller) framework.  Widgets defined in the back end are automatically synchronized with generic Backbone.js models in the front end.  The traitlets are added to the front end instance automatically on first state push.  The `_view_name` trait that you defined earlier is used by the widget framework to create the corresponding Backbone.js view and link that view to the model."
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:365]: Perhaps add some more explanations:

```suggestion
    "The IPython widget framework front end relies heavily on [Backbone.js](http://backbonejs.org/).  Backbone.js is an MVC (model view controller) framework.  Widgets defined in the back end are automatically synchronized with Backbone.js `Model` in the front end. Each front end `Model` handles the widget data and state, and can have any number of associate `View`s. In the context of a widget the `Views` are what render objects for the user to interact with, and the Model handles communication with the Python objects.\n",
    "\n",
    "On the first state push from python the synced traitlets are added automatically. The `_view_name` trait that you defined earlier is used by the widget framework to create the corresponding Backbone.js view and link that view to the model.\n"
```


May also be worthwhile to discuss the defaults. Though I'm not actually 100% how/when they are important. This may come up for a reader especially in the context of `_view_name` where the text claims it is set by python, but the code seems to imply that it also comes from `defaults`

File: docs/source/examples/Widget Custom.ipynb [hunk 11]
```diff
@@ -0,0 +1,785 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The widget framework is built on top of the Comm framework (short for communication).  The Comm framework is a framework that allows the kernel to send/receive JSON messages to/from the front end (as seen below).\n",
+    "\n",
+    "![Widget layer](images/WidgetArch.png)\n",
+    "\n",
+    "To create a custom widget, you need to define the widget both in the browser and in the Python kernel.\n",
+    "\n",
+    "This tutorial shows how to build a simple email widget using the TypeScript widget cookiecutter: https://github.com/jupyter-widgets/widget-ts-cookiecutter"
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:33]: I'd put the screenshot of the completed widget here as well. So that people have a sense of what they'll get at the end of the tutorial

File: docs/source/examples/Widget Custom.ipynb [hunk 12]
```diff
@@ -0,0 +1,785 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The widget framework is built on top of the Comm framework (short for communication).  The Comm framework is a framework that allows the kernel to send/receive JSON messages to/from the front end (as seen below).\n",
+    "\n",
+    "![Widget layer](images/WidgetArch.png)\n",
+    "\n",
+    "To create a custom widget, you need to define the widget both in the browser and in the Python kernel.\n",
```
Comment (ianhi) [inline docs/source/examples/Widget Custom.ipynb:31]: This section is nice, but feels a bit out of place at the top of the page. What do you think about moving it to the bottom? Then it could be a "How did this all work" section.

Comment (ianhi) [reply docs/source/examples/Widget Custom.ipynb:31]: Perhaps also link to the widget low level notebook that goes into far more detail.

**Review by ianhi — COMMENTED** [review]
This is looking great! I left a comments on a few small things which can mostly done in subsequent PRs. I think the only things that really need addressing are contents of some of the code blocks. Following along I found that I was misled into deleting things, or never ended up defining a necessary variable a few times.

Other than that a great read, I learned a few things along the way :)

File: docs/source/examples/Widget Custom.ipynb (hunk 6 above)
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:451]: Good point, adding that :+1:

File: docs/source/examples/Widget Custom.ipynb (hunk 7 above)
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:461]: We can add the lines for completeness yes.

File: docs/source/examples/Widget Custom.ipynb (hunk 12 above)
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:31]: Moving this block a bit below.

File: docs/source/examples/Widget Custom.ipynb (hunk 11 above)
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:33]: Sounds good, thanks!

**Comment by jtpio** [comment]
Thanks @ianhi for the comments, I've pushed changes to the branch which should be available on the RTD preview now too: https://ipywidgets--2919.org.readthedocs.build/en/2919/examples/Widget%20Custom.html

File: docs/source/examples/Widget Custom.ipynb [hunk 13]
```diff
@@ -0,0 +1,785 @@
...
+    "- Long\n",
+    "- Set\n",
+    "- TCPAddress\n",
+    "- Tuple\n",
+    "- Type\n",
+    "- Unicode\n",
+    "- Union\n",
+    "\n",
+    "\n",
+    "Not all of these traitlets can be synchronized across the network, only the JSON-able traits and Widget instances will be synchronized."
```
[the remark this answers is not in the document]
Comment (jtpio) [reply docs/source/examples/Widget Custom.ipynb:348]: We can probably go through the list and check them all with a call to `json.dumps()` to know which ones are serializable.

**Comment by jtpio** [comment]
We could probably iterate more on this tutorial in separate PRs, to cover the following points:

- Update instructions to JupyterLab 3.0
- List non JSON serializable trait types

**Review by ianhi — APPROVED** [review]
This is fantastic :tada: thanks!

> We could probably iterate more on this tutorial in separate PRs, to cover the following points:

:+1:
